plugin/src/main/java/at/helpch/chatchat/util/MessageProcessor.java

```diff
@@ -42,33 +42,47 @@ public static void process(
-            final var canMentionEveryone = user.player().hasPermission(MENTION_EVERYONE_PERMISSION);
-            final var prefix = plugin.configManager().settings().mentionPrefix();
-            final var sound = plugin.configManager().settings().mentionSound();
-            for (final String word: message.split(" ")) {
-                if (!word.startsWith(prefix)) continue;
-                final var name = word.substring(1);
-
-                if (canMentionEveryone && (name.equals("everyone") || name.equals("here"))) {
-                    channel.playSound(sound);
-                    break;
-                }
-
-                final var player = Bukkit.getPlayer(name);
-                if (player == null) continue;
-                plugin.usersHolder().getUser(player).playSound(sound);
-            }
+        final AtomicBoolean pingEveryone = new AtomicBoolean(false);
+        final var mentionPrefix = plugin.configManager().settings().mentionPrefix();
 
+        // FIXME - tidy up
+        if (user.player().hasPermission(MENTION_EVERYONE_PERMISSION)) {
+            chatEvent.message(chatEvent.message().replaceText(builder -> builder
+                    .match(mentionPrefix + "(here|everyone)")
+                    .replacement((result, ignored) -> {
+                                pingEveryone.set(true);
+                                return Component.text(result.group()).color(NamedTextColor.YELLOW);
+                            }
+                    )));
         }
 
         final var oldChannel = user.channel();
         user.channel(channel);
-        chatEvent.channel().sendMessage(chatEvent.user(), FormatUtils.parseFormat(
-            chatEvent.format(),
-            user.player(),
-            chatEvent.message()
-        ));
-        user.channel(oldChannel);
+        final var component = FormatUtils.parseFormat(
+                chatEvent.format(),
+                user.player(),
+                chatEvent.message()
+        );
+
+        // FIXME - this is messy. make it not
+        final var canMention = user.player().hasPermission(MENTION_PERMISSION);
+        for (final var target : chatEvent.channel().targets(chatEvent.user())) {
+            var userSpecificComponent = component;
+            if (canMention && target instanceof ChatUser) {
+                final var name = ((ChatUser) target).player().getName();
+                userSpecificComponent = component.replaceText(builder ->
+                        builder.matchLiteral(mentionPrefix + name)
+                                .replacement((result, ignored) -> {
+                                    target.playSound(plugin.configManager().settings().mentionSound());
+                                    return Component.text(result.group()).color(NamedTextColor.YELLOW);
+                                })
+                );
+            }
+            target.sendMessage(user, userSpecificComponent);
```

Comment:
Shouldn't we still have a custom sendMessage method in the channel? That's what I was struggling with. Putting the sendMessage method in channels.

Reply:
Pretty sure that was the reason to move to adventure. bcz it was making it easier. I mean we can make it not adventure but still.

Reply:
not sure if it's really appropriate to be honest, given that each channel is implemented differently

Reply:
Well exactly. So we can implement a different sendmessage per channel.